nupic/research/frameworks/dendrites/modules/dendritic_mlp.py

```diff
@@ -0,0 +1,225 @@
+                    param.requires_grad = False
+
+    def forward(self, x, context):
+        output = self.dend1(x, context=context)
+        output = self.kw1(output) if self.kw else output
+
+        output = self.dend2(output, context=context)
+        output = self.kw2(output) if self.kw else output
+
+        output = self.classifier(output)
+        return output
+
+    # ------ Weight initialization functions
+    def _init_sparse_weights(self, m, input_sparsity):
+        """
+        Modified Kaiming weight initialization that considers input sparsity and weight
+        sparsity.
+        """
+        input_density = 1.0 - input_sparsity
+        weight_density = 1.0 - m.sparsity
+        _, fan_in = m.module.weight.size()
+        bound = 1.0 / np.sqrt(input_density * weight_density * fan_in)
+        nn.init.uniform_(m.module.weight, -bound, bound)
+        m.apply(rezero_weights)
+
+    def _init_sparse_dendrites(self, m, input_sparsity):
+        """
+        Modified Kaiming initialization for dendrites segments that consider input
+        sparsity and dendritic weight sparsity.
+        """
+        # Assume `m` is an instance of `DendriticLayerBase`
+        input_density = 1.0 - input_sparsity
+        weight_density = 1.0 - m.segments.sparsity
+        fan_in = m.segments.dim_context
+        bound = 1.0 / np.sqrt(input_density * weight_density * fan_in)
+        nn.init.uniform_(m.segments.weights, -bound, bound)
+        m.apply(rezero_weights)
+
+    def hardcode_dendritic_weights(self, context_vectors, init):
+        """
+        Set up specific weights for each dendritic segment based on the value of init.
+
+        if init == "overlapping":
+            We hardcode the weights of dendrites such that each context selects 5% of
+            hidden units to become active and form a subnetwork. Hidden units are
+            sampled with replacement, hence subnetworks can overlap. Any context/task
+            which does not use a particular hidden unit will cause it to turn off, as
+            the unit's other segment(s) have -1 in all entries and will yield an
+            extremely small dendritic activation.
+
+        otherwise if init == "non_overlapping":
+            We hardcode the weights of dendrites such that each unit recognizes a single
+            random context vector. The first dendritic segment is initialized to contain
+            positive weights from that context vector. The other segment(s) ensure that
+            the unit is turned off for any other context - they contain negative weights
+            for all other weights.
+
+        :param context_vectors:
+        :param init: a string "overlapping" or "non_overlapping"
+        """
+        self._hardcode_dendritic_weights(self.dend1.segments, context_vectors, init)
+        self._hardcode_dendritic_weights(self.dend2.segments, context_vectors, init)
+
+    def _hardcode_dendritic_weights(self, dendrite_segments, context_vectors, init):
```

Comment:
`_hardcode_dendritic_weights` doesn't use `self`, so once again it's unnecessary to have it as a method rather than a separate function. We should collapse hardcoding into a single method, or move this 2nd method to a stand-alone function.

Reply:
Same answer as before. I'll make it into a static method. I don't think we can make it a single method because it's called twice in hardcode_dendritic_weights()